# breast cancer wisconsin prediction using knn
dataset http://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+(Original)

### Attribute Information: (class attribute has been moved to last column)

   #  Attribute                     Domain
   -- -----------------------------------------
   1. Sample code number            id number
   2. Clump Thickness               1 - 10
   3. Uniformity of Cell Size       1 - 10
   4. Uniformity of Cell Shape      1 - 10
   5. Marginal Adhesion             1 - 10
   6. Single Epithelial Cell Size   1 - 10
   7. Bare Nuclei                   1 - 10
   8. Bland Chromatin               1 - 10
   9. Normal Nucleoli               1 - 10
   10. Mitoses                       1 - 10
   11. Class:                        (2 for benign, 4 for malignant)


Missing attribute values: 16

 There are 16 instances in Groups 1 to 6 that contain a single missing 
   (i.e., unavailable) attribute value, now denoted by "?".  

### 1 DataPrepration

In [118]:
#import library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.cross_validation import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.cross_validation import cross_val_score
from collections import Counter
from sklearn.metrics import accuracy_score
from sklearn import cross_validation

In [119]:
#reading data.csv naming columns 'id','clump_thickness','unif_cell_size','unif_cell_shape,marg_adhesion','single_epith_cell_size','bare_nuclei','bland_chrom','norm_nucleoli','mitoses','class'
df=pd.read_csv("data.csv",names=['id','clump_thickness','unif_cell_size','unif_cell_shape,marg_adhesion','single_epith_cell_size','bare_nuclei','bland_chrom','norm_nucleoli','mitoses','class'])
#droping unnecessary column id 
df.drop(["id"],axis=1,inplace=True)
df.shape

(699, 9)

In [120]:
df.replace('?',0,inplace=True)

In [121]:
df["class"].value_counts()

2    458
4    241
Name: class, dtype: int64

# Observation
1.there are total 699 patient<br>
2.out of 699 patient 458 are having benign and 241 are having malignant

### 2.High level statistics

In [122]:
df.describe()

clump_thickness  unif_cell_size  unif_cell_shape,marg_adhesion  \
count       699.000000      699.000000                     699.000000   
mean          3.134478        3.207439                       2.806867   
std           3.051459        2.971913                       2.855379   
min           1.000000        1.000000                       1.000000   
25%           1.000000        1.000000                       1.000000   
50%           1.000000        1.000000                       1.000000   
75%           5.000000        5.000000                       4.000000   
max          10.000000       10.000000                      10.000000   

       single_epith_cell_size  bland_chrom  norm_nucleoli     mitoses  \
count              699.000000   699.000000     699.000000  699.000000   
mean                 3.216023     3.437768       2.866953    1.589413   
std                  2.214300     2.438364       3.053634    1.715078   
min                  1.000000     1.000000       1.000000    1.000000   
25%                  2.000000     2.000000       1.000000    1.000000   
50%                  2.000000     3.000000       1.000000    1.000000   
75%                  4.000000     5.000000       4.000000    1.000000   
max                 10.000000    10.000000      10.000000   10.000000   

            class  
count  699.000000  
mean     2.689557  
std      0.951273  
min      2.000000  
25%      2.000000  
50%      2.000000  
75%      4.000000  
max      4.000000

### 3.k nearest neighbor

In [123]:
#putting all class in cls
cls=df['class']
df.drop(['class'],axis=1,inplace=True)

In [124]:
#spliting total data point 30% test and 70% test data
x_tr,x_test,y_tr,y_test=cross_validation.train_test_split(df,cls,test_size=0.3,random_state=0)
score=[]
#looping value from 1 to 20 putting to find optimal value of K
for x in range(1,20):
    knn=KNeighborsClassifier(n_neighbors=x)
    #using 10-Fold cross validation
    scores=cross_val_score(knn,x_tr,y_tr,cv=10,scoring='accuracy')
    score.append(np.mean(scores)*100)

for idx, val in enumerate(score):
    print(idx, val)

0 95.10493197278912
1 93.85510204081632
2 96.31717687074828
3 96.10884353741497
4 96.32125850340137
5 96.52959183673471
6 96.72534013605443
7 96.10884353741497
8 96.10884353741497
9 95.70051020408164
10 95.90884353741498
11 95.90884353741498
12 96.32125850340137
13 95.90884353741498
14 96.11717687074831
15 96.32551020408164
16 96.11717687074831
17 96.11717687074831
18 96.32551020408164


# Observation
maximum accuracy is 96 when K value is 2-7

In [108]:
#training Knn k=5
knn_opitmal=KNeighborsClassifier(5)
knn_opitmal.fit(df,cls)
predict=knn_opitmal.predict(x_test)
acc = accuracy_score(y_test, predict) * 100
print(acc)

94.76190476190476


# Conclusion
we can predict breast cancer with the accuracy of 94%